src/coordinator/executor/transform/cache.go

```diff
@@ -70,3 +71,93 @@ func (c *BlockCache) Get(key parser.NodeID) (block.Block, bool) {
 	b, ok := c.blocks[key]
 	return b, ok
 }
+
+type cacheTime int64
+
+func fromTime(t time.Time) cacheTime {
+	return cacheTime(t.UnixNano())
+}
+
+func (c cacheTime) toTime() time.Time {
+	return time.Unix(0, int64(c))
+}
+
+// TimeCache is used to cache blocks over time. It also keeps track of blocks which have been processed
+type TimeCache struct {
+	blocks          map[cacheTime]block.Block
+	processedBlocks map[cacheTime]bool
+	mu              sync.Mutex
+}
+
+// NewTimeCache creates a new TimeCache
+func NewTimeCache() *TimeCache {
+	return &TimeCache{
+		blocks:          make(map[cacheTime]block.Block),
+		processedBlocks: make(map[cacheTime]bool),
+	}
+}
+
+// Add the block to the cache, errors out if block already exists
+func (c *TimeCache) Add(key time.Time, b block.Block) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	_, ok := c.blocks[fromTime(key)]
+	if ok {
+		return errors.New("block already exists")
```

Comment:
Does this need to error? If it's already in there, isn't that fine? If you need to, can we just return a `bool` instead?

Reply:
a block already there is probably a bug so an error helps us identify it